### DISCUSSION TITLE
created a hair dryer in replicad

### DISCUSSION THREAD

**Discussion by raydeleu** [sgenoud/replicad#10] [Show and tell]
I tried to model a hairdryer in Replicad. 

![hairdryer-rc](https://user-images.githubusercontent.com/38007983/171355931-4783ae1f-e702-4c87-b047-b94b0c7498e6.png)

The shape is interesting because it holds the middle ground between an organic shape and a hard surface shape. So modellng in Blender is not easy, but it is also challenging for most CAD-like programs. Especially creating the fillet between the fanhouse, the nozzle and the handle is difficult. In Replicad I was not able to create the fillet between the handle and the fanhouse. The code below contains two approaches to select the edges, but in both cases the model crashes and creates a kind of blob in the Visualizer. 

```` javascript
const main = (
{ 
    sketchCircle,
    sketchRectangle,
    EdgeFinder
}
) => {

// parameters
let fanRadius       = 30;
let fanHeight       = 30;
let fanhousingThickness = 1.5;
let fanCutoutRadius = 20;
let fanCutoutDepth  = 5 ; 
let fanRounding     = 8;
let lidRadius       = 19;
let lidThickness    = 3; 
let lidRounding     = 2;
let stemRadius      = 5;
let outletLength    = 60;
let outletWidth     = 18;
let outletHeight    = 30; 
let outletRounding  = 5;

let outletThickness = 1.45;
let handleLength    = 80;
let handleWidth     = 25;
let handleHeight    = 16; 
let handleRounding  = 5;
let handleThickness = 1.45;
let handleBottomRound = 2;
let buttonLength    = 8;
let buttonWidth     = 3;
let buttonHeight    = 6;
let buttonRounding  = 1.4;
let recesDepth      = 20;
let recesWidth      = 10;
let recesHeight     = 10;
let recesRounding   = 2; 
let handleJunctionRound = 5;
let outletJunctionRound = 5; 

// fanhousing
let fanhousing = sketchCircle(fanRadius).extrude(fanHeight).fillet(fanRounding);
let cutout     = sketchCircle(fanCutoutRadius).extrude(fanCutoutDepth+10).translateZ(fanHeight-fanCutoutDepth);
fanhousing = fanhousing.cut(cutout);

// lid
let lid = sketchCircle(lidRadius).extrude(lidThickness);
lid = lid.fillet(lidRounding, (e) => e.ofCurveType("CIRCLE").inPlane("XY",lidThickness) );
lid = lid.translateZ(fanHeight-lidThickness);
let lidstem = sketchCircle(stemRadius).extrude(fanHeight)
lid = lid.fuse(lidstem);

// outlet
let outletBase  = (fanHeight - outletWidth)/2;
let outlet      = sketchRectangle(outletHeight,outletLength).extrude(outletWidth);
outlet = outlet.translate([(-fanRadius+(outletHeight/2.0)),-outletLength/2,outletBase]);
outlet  = outlet.fillet(outletRounding, (e) => e.inDirection("Y"));

fanhousing = fanhousing.fuse(outlet);
let corner1 = [0, -outletLength,-fanHeight/2];
let corner2 = [-fanRadius,-fanRadius-5,fanHeight/2];
fanhousing = fanhousing.fillet(outletJunctionRound, (e) => e.inDirection("Z").not( (f) => f.inBox(corner1,corner2)))

// handle
let handleBase = (fanHeight - handleHeight)/2;
let handle = sketchRectangle(handleLength,handleWidth).extrude(handleHeight)
handle = handle.translate([handleLength/2,fanRadius-handleWidth/2,handleBase])
handle = handle.fillet(handleRounding, (e) => e.inDirection("X"))
handle = handle.fillet(handleBottomRound, (e) => e.inPlane("YZ",handleLength))
fanhousing = fanhousing.fuse(handle);
let corner3 = [fanRadius+2,fanRadius+2,-fanHeight/2+1]
let corner4 = [0 ,0 ,fanHeight/2-1]
//fanhousing = fanhousing.fillet(handleJunctionRound, (e) => e.inBox(corner3,corner4));
// fanhousing = fanhousing.fillet(handleJunctionRound, (e) => e.inDirection("Z"));

// create a shell with thickness of the combined shape
fanhousing = fanhousing.shell(-fanhousingThickness, (f) => f.inPlane("XZ",[-fanRadius,-outletLength,0]))

// add the lid into the reces of the fanhousing
fanhousing = fanhousing.fuse(lid);
return fanhousing;
}
````

**Comment by lf94** [comment]
(I've subscribed to discussions here so I got the email 🙂)

It's probably because the fillet is too large and is overextending to something. I've encountered similar errors in CADQuery (which use the same underlying engine)...

Try with an ever increasing chamfer first to see where it overextends, and then switch to fillet...

This issue is why I'm torn between a few different Code CAD today. Doing organic or non-organic surfaces has no nice inbetween. It's why I'm working on my own Code CAD but really don't have the time currently to push it forward...

**Comment by raydeleu** [comment]
I have tried your suggestion but it did not work. I know that the shape should be possible, as I modelled it earlier in CascadeStudio. I know that in earlier versions I had to tweak the dimensions a bit to make it work, but when I tried it again today it worked perfectly. 

![hairdryer-cs2](https://user-images.githubusercontent.com/38007983/171484847-065a62e4-1d16-477c-b9cf-de3a20d24d98.png)

Note that the cut-out in the nozzle looks a bit weird as I shifted the inner offset of the nozzle before subtracting it from the shape to remove the outlet face. 

![hairdryer-cs3](https://user-images.githubusercontent.com/38007983/171485779-1705c712-127f-418c-9af5-c56a4d3d8517.png)

A very similar model in OnShape: 

![hairdryer-onshape](https://user-images.githubusercontent.com/38007983/171484978-d4b4f306-dd33-49f3-94f7-9ab6d11f1fdd.png)

So I assume that the kernel should be able to calculate the shape. Therefore I guess that the selection of the edges does not work as expected.

**Reply by sgenoud** [reply]
Make sure you have reloaded the page to have the latest version of the app - once you do it should work (also with your method)!

**Comment by raydeleu** [comment]
Success! I don't know what magic was performed behind the scenes, but I indeed got a model. 

![hairdryer-rc2](https://user-images.githubusercontent.com/38007983/171499045-d512b776-79b6-4dca-9869-a64a9c6503b8.png)

While writing this reaction I realized why the method where I select a vertical edge does not work. At this point in the model creation there is another vertical edge that I excluded from the selection of edges earlier. So the correct edge selection now looks like this: 


````  javascript
//  code to fillet the junction between the nozzle and the fan housing
fanhousing = fanhousing.fillet(outletJunctionRound, (e) => e.inDirection("Z").not( (f) => f.inBox(corner1,corner2)))

// code to fillet the junction between the handle and the fanhousing 

fanhousing = fanhousing.fillet(handleJunctionRound, (e) => e.inBox(corner3,corner4));   // version 1
//  fanhousing = fanhousing.fillet(handleJunctionRound, 
//       (e) => e.inDirection("Z").not( (f) => f.inBox(corner1,corner2)));        // version 2

````

Both version 1 and version 2 work in the most recent version of the app.
